ipq50xx: package: add glinet-uboot-scr package

TheRootED24 opened the pull request:
** glinet-uboot-scr package

* This package contians the required uboot scr file required to enable uboot webui firmware upgrades and recovery options. These scripts, like bdf files are device specific. This seems to be the first device with these requirements, but i suspect more to be discovered now that that method is known and proven. This is an option not a requirement for the GL.iNet GL-b3000 device port. It must be included in the build to have any effect, as the title and description explain in detail

robimarko commented:
Nice, its just the QSDK installation script for U-Boot

TheRootED24 commented:
Yes,  the oem (gl-inet in my case) bakes it into the image. Its not complex to deal with,but makes a real sore spot on a new device pr. Do you find this to be a suitable way to deal with it ?   

My first thought was to just stash it someplace, like the scripts directory for example ,  but it sticks out like a boot script in there.  So my next thought was a script that would consume it (cloak it as a regular build script) and echo it out when called, but this seemed icky to me as well.  

So will all things considered,  I figured this would be a better way to deal with it.  By using webui upgrades to flash openwrt from oem, and uboot with the oem .img file to revert back to stock from openwrt , the port is not dependent on uboot upgrades,  so this ugly script doesn't need to reside in the build system. My thoughts anyway..

With that said, If you can think of a better way, or would like to do it differently, i have no objections what so ever. Being this is an issue which hasn't really been discussed previously,  I haven't received much advice in regards to the best way to handle this. I am open to any and all suggestions you may have.

Rondom commented:
I think it is better to generate this script. It is aurogenerated in QSDK. We probably need to write our own. In QSDK, there are two versions of the script, one is proprietary and processes some XML files *shudder* describing each board and then there is another one that is inside their U-Boot repo which might be GPL. I have not looked at the latter one and whether it is any saner.

TheRootED24 commented:
Sure i have no problem with this. This was my first proposal, but i received friction about it.  Could you pop by my pr for the glinet-gl-b3000, I need to come up with a solution to move my pr along. I figured this would be the fastest way to get my device added, if there was no impact to the build system and i toted the weight of maintaining the package

pop in and have a look if you would, i could use some good advice to get it across the line.

[glinet_gl-b3000 pr]( https://github.com/openwrt/openwrt/pull/17903)

I will leave this pr until something is decided moving forward.

TheRootED24 committed:
ipq50xx: package: add glinet-uboot-scr pacakge

** glinet-uboot-scr package

* This package contians the required uboot scr file
  required to enable uboot webui firmware upgrades and
  recovery options. These scripts, like bdf files are
  device specific. This seems to be the first device with
  these requirements, but i suspect more to be discovered
  now that that method is known and proven. This is an option
  not a requirement for the GL.iNet GL-b3000 device port. It must
  be included in the build to have any effect, as the title as
  description explain in detail

This has been tested thorougly in development of the GL.iNet B3000.
The complete discussion, tests, and results can been seen in the current pr
here --> openwrt#17903 (comment)

Signed-off-by: Scott Mercer <TheRootEd24@gmail.com>

robimarko commented:
> Sure i have no problem with this. This was my first proposal, but i received friction about it. Could you pop by my pr for the glinet-gl-b3000, I need to come up with a solution to move my pr along. I figured this would be the fastest way to get my device added, if there was no impact to the build system and i toted the weight of maintaining the package

Who NACK-ed building the script directly in OpenWrt?
Cause we are already building those scripts with mkimage for mvebu and like, seems like overkill to have an external package just for this when its just one U-Boot script

TheRootED24 commented:
>Cause we are already building those scripts with mkimage for mvebu and like

Ok,  I was not aware of this. you're creating the script with mkimage ? or just included them in the .its ?  From what i see,  looking at the scripts from the gl-b3000 and the gl-x20000 , the scripts are different, so we will need a place to put them unless you are in fact generating them on the fly.

perhaps an example will clarify things, can you point me there ?

robimarko commented:
Look at:
https://github.com/openwrt/openwrt/blob/main/target/linux/mvebu/image/Makefile#L14

This calls mkimage to make the final scripts in the binary form for U-boot from:
https://github.com/openwrt/openwrt/tree/main/target/linux/mvebu/image

I dont see why we need an external package, cause in your case it just hosts an finished script?

TheRootED24 commented:
ok, so basically just include the boot scripts in the respective image directories alongside the makefiles. simple enough.

The friction I spoke of was in the pr for the gl-b3000, in regards as to where to put the script/s, no one could decide on the proper location or the proper way in which to handle them. But since its already being done in this way, I will just follow suite and adjust the pr for the gl-b3000 accordingly.

I will wait for comment before i close this pr  

thx rob

TheRootED24 commented:
for the sake of clarity,  here is my current build recipe. 

It checks for the --with-uboot-scr flag and also checks if the script exists in .. currently where the package installs it in the staging dir , as seen here

```
$(eval GL_INCLUDE := $(STAGING_DIR)/usr/include/glinet-uboot-scr) // <-- target directory
$(if $(wildcard $(GL_INCLUDE)/*),$(CP) $(GL_INCLUDE)/* $(KDIR_TMP)/) \ // <--- if script exist cp to kdir_tmp
$(if $(and $(wildcard $(GL_SCRIPT)),$(findstring --with-uboot-scr,$(word 1,$(1)))),-s $(GL_SCRIPT)) \  <--- check for flag & scr

```

full recipe
```
define Build/gl-factory
	$(eval GL_NAME := $(subst $(comma),_,$(word 1,$(SUPPORTED_DEVICES))))
	$(eval GL_INCLUDE := $(STAGING_DIR)/usr/include/glinet-uboot-scr)
	$(eval GL_SCRIPT := $(GL_INCLUDE)/$(GL_NAME).scr)
	$(eval GL_IMGK := $(KDIR_TMP)/$(DEVICE_IMG_PREFIX)-squashfs-factory.img)
	$(eval GL_ITS := $(KDIR_TMP)/$(GL_NAME).its)
	$(if $(wildcard $(GL_INCLUDE)/*),$(CP) $(GL_INCLUDE)/* $(KDIR_TMP)/) \

	$(TOPDIR)/scripts/mkits-glinet.sh \
		$(if $(and $(wildcard $(GL_SCRIPT)),$(findstring --with-uboot-scr,$(word 1,$(1)))),-s $(GL_SCRIPT)) \
		-f $(GL_IMGK) \
		-o $(GL_ITS) \

	PATH=$(LINUX_DIR)/scripts/dtc:$(PATH) mkimage -f \
		$(GL_ITS) \
		$(GL_IMGK)
endef

```
so in my case, i can just add my boot script/s to the image directory and change the target path in my build recipe.


thinking foward .. 

I am not sure how many supported devices use these scripts currently. But perhaps it would be cleaner to store them all in a single package like ipq-wifi does with the bdf's,  if there is ...or when then,  enough of them to justify it ??

that was my original intention with the separate package idea. However, I didn't structure it in that way because i was unaware that there were already supported devices using them. But perhaps now or in the future we can consolidate them in this way ...or it least in a single directory within the build system.
